src/components/tl/ucp/tl_ucp_coll.h

```diff
@@ -188,7 +189,11 @@ ucc_tl_ucp_init_task(ucc_base_coll_args_t *coll_args, ucc_base_team_t *team)
 
     ucc_coll_task_init(&task->super, coll_args, team);
     tl_team->seq_num = (tl_team->seq_num + 1) % UCC_TL_UCP_MAX_COLL_TAG;
```

Comment:
Come to thinking about one scenario that one calls pt2pt without setting the tag. So those two ranks who are involved in pt2pt will use and increase the `seq_num`, while other ranks won't. This seems to cause tag mismatch in next set of collectives or pt2pt from different pair.

Just a few thoughts for brainstorming:
1) move this increment to the `else` block below (i.e., only increase seq_num if tag is not used), this probably not enough
2) have two internal seq_num with different ranges, one for pt2pt and one for collective to avoid collision. And plus 1) to avoid mismatch
3) force/assume all ranks to be involved during init even for pt2pt (sounds unnecessary) 
4) force pt2pt to always provide tag (even a dummy one) and only use internal seq_num for collective

Reply:
@kingchc and I had a conversation offline about this, but to summarize here:

> 1. move this increment to the `else` block below (i.e., only increase seq_num if tag is not used), this probably not enough

That doesn't solve the problem unless you pair with 4) below and require the user to always provide a tag when using pt2pt (which doesn't necessarily make sense if you're using something like the NCCL TL where tags aren't supported anyway).

> 2. have two internal seq_num with different ranges, one for pt2pt and one for collective to avoid collision. And plus 1) to avoid mismatch

This would solve the problem, but only if you keep track of a sequence number per-pair, which gets pretty unwieldy. My preference would be to say that if the user doesn't specify a tag, we just use a reserved value and all messages get sent with that tag (after adding the high-order bit to indicate that it's a pt2pt message). That way it just acts like an untagged message (in the sense that it can be matched in any order), but it doesn't conflict with any of the collective messages.

> 3. force/assume all ranks to be involved during init even for pt2pt (sounds unnecessary)

😄 

> 4. force pt2pt to always provide tag (even a dummy one) and only use internal seq_num for collective

This is the second best option in my mind in terms of keeping a clean interface. However, it does add a burden on the user that might be a little weird (see above). My preference would be for 2 and I'll update the PR to reflect that.